app/src/main/java/space/taran/arknavigator/mvp/model/repo/PreviewsRepo.kt

```diff
@@ -0,0 +1,198 @@
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import space.taran.arknavigator.mvp.model.dao.ResourceId
+import space.taran.arknavigator.ui.App
+import space.taran.arknavigator.ui.fragments.utils.Preview
+import space.taran.arknavigator.utils.*
+import space.taran.arknavigator.utils.extensions.autoDisposeScope
+import space.taran.arknavigator.utils.extensions.makeGone
+import space.taran.arknavigator.utils.extensions.textOrGone
+import java.io.File
+import java.io.FileOutputStream
+import java.nio.file.Files
+import java.nio.file.Path
+import com.shockwave.pdfium.PdfiumCore
+
+class PreviewsRepo {
+
+    fun providePreview(path: Path): Preview {
+        if (Files.isDirectory(path)) {
+            return Preview(filePath = path, isFolder = true)
+        }
+        return createPreview(path)
+    }
+
+    private fun createPreview(file: Path): Preview {
+        return if (isPDF(file)) getPdfPreview(file)
+        else Preview(filePath = file, fileExtension = extension(file))
+    }
+
+    private fun getPdfPreview(file: Path): Preview {
+
+        val encodedID = Base64.encodeToString(file.toString().toByteArray(), Base64.URL_SAFE)
+        val savedPreviews = getSavedPdfPreviewsDecoded()
+
+        return if (savedPreviews?.contains(encodedID) == true) {
+            Preview(
+                filePath = getPdfPreviewPathByID(encodedID),
+                fileExtension = extension(file)
+            )
+        } else Preview(
+            filePath = file,
+            fileExtension = extension(file)
+        )
+    }
+
+    fun generatePdfPreviewsForMeta(metaByPath: MutableMap<Path, ResourceMeta>) {
+        val previewsFolder = getPdfPreviewsFolder()
+        val savedPreviews = getSavedPdfPreviewsDecoded()
+
+        metaByPath.forEach {
+            val path = it.key
+            var out: FileOutputStream? = null
+
+            val encodedID = Base64.encodeToString(path.toString().toByteArray(), Base64.URL_SAFE)
+
+            if (savedPreviews == null || !savedPreviews.contains(encodedID)) {
+                if (isPDF(path) && it.value.size / MEGABYTE >= 5) {
+                    try {
+                        if (!previewsFolder.exists()) previewsFolder.mkdirs()
+
+                        val file = File(previewsFolder, "$encodedID.png")
+                        out = FileOutputStream(file)
+                        createPdfPreview(path)
+                            ?.compress(Bitmap.CompressFormat.PNG, 100, out)
+                    } catch (e: Exception) {
+                    } finally {
+                        try {
+                            out?.close()
+                        } catch (e: Exception) {
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fun loadPreview(
+        targetView: ImageView,
+        preview: Preview,
+        extraMeta: ResourceMetaExtra?,
+        centerCrop: Boolean = false
+    ) {
```

Comment:
In the case of `fun loadExtraMeta` I was able to remake the implementation so that it doesn't have to work with views.
In this case, I'm not sure how to do this. The whole point of moving this method to PreviewsRepo was to remove this heavy logic from adapters, that just have a job of displaying the data.

The issue here is that there are a lot of cases, and it's hard to encapsulate them all into some action. For example, for gifs we call methods `loadGif` or `loadGifThumbnailWithPlaceHolder` depending on the type of image. PDF's launch a coroutine in the context of the view to creade a bitmap of preview, and others may call `loadCroppedImageWithPlaceHolder` or `loadImageWithTransition` depending on other cases. How can we remake this, but keeping the business logic outside of adapters?

Reply:
Maybe create `PreviewHelper` or `PreviewLoader` in `utils`, that can resolve the preview and load it into the view?

Reply:
This method should be decomposed better. E.g. we don't need to generate previews during loading. We've already decided to generate them during indexing. After they are generated, it's only left to load them and this doesn't need the state at all. By the way, `PreviewsRepo` doesn't have any state (fields), all methods of it are static, we don't need it even to be `class`. Also, the name is confusing, "repo" usually is interface to some storage.